python/ray/dag/dag_node_operation.py

```diff
@@ -0,0 +1,102 @@
+from functools import total_ordering
+from enum import Enum
+from ray.util.annotations import DeveloperAPI
+import ray
+
+
+@DeveloperAPI
+class DAGNodeOperationType(Enum):
+    """
+    There are three types of operations that a DAG node can perform:
+    1. READ: Read from an input channel.
+    2. COMPUTE: Execute the method corresponding to the node.
+    3. WRITE: Write to an output channel.
+    """
+
+    READ = "READ"
+    COMPUTE = "COMPUTE"
+    WRITE = "WRITE"
+
+
+@DeveloperAPI
+class DAGNodeOperation:
+    def __init__(
+        self,
+        idx: int,
+        operation_type: DAGNodeOperationType,
+    ):
+        """
+        Args:
+            idx: The index of the task that this operation belongs to
+                in the actor's ExecutableTask list. The index is not
+                the same as bind_index, but there are positive correlations
+                between the two.
+            operation_type: The type of operation to perform.
+        """
+        self.idx = idx
+        self.type = operation_type
+
+
+@DeveloperAPI
+@total_ordering
+class DAGOperationGraphNode:
+    def __init__(
+        self,
+        operation: DAGNodeOperation,
+        idx: int,
+        actor_handle: "ray.actor.ActorHandle",
+        requires_nccl: bool,
+    ):
+        """
+        DAGOperationGraphNode represents a node in the DAG operation graph.
+        It contains information about the node's in-degree, out-degree, edges,
+        and the operation it performs.
+
+        Args:
+            operation: The operation that this node performs. The operation
+                can be a READ, COMPUTE, or WRITE operation.
+            idx: A unique index into the original DAG.
```

Comment:
this is really confusing with idx for `DAGNodeOperation`. Can you explain to me how these 2 idx are differnet?

Reply:
Maybe I should use `global_idx` and `local_idx` instead.

* `global_idx` (A unique index into the original DAG.): This index can be used as the key in `idx_to_tasks` to retrieve a `CompiledTask`. `CompiledTask` is used to associate `ActorHandle`, `requires_nccl`, and `downstream_node_idxs`.

* `local_idx` (`operation.idx`): This index is the index of the `ExecutableTask` to which this operation belongs in the `executable_tasks` list (a per-actor list), which is sorted by `bind_index`. Note that this is a bit different from `bind_index`, although `local_idx` has a positive correlation with `bind_index`.

`local_idx` is possible to be different from `bind_index` because an actor can be reused across different `CompiledDAG`.

Reply:
Does using `global_idx` and `local_idx` make sense to you?

Reply:
is it possible to remove local_index and only keep bind_index, or is it difficult? If it is difficult, global & local index sound good

Reply:
Sorry, I am still confused by the explanation of local_index. Can you give an example of when it is different from bind_index?

Reply:
https://github.com/ray-project/ray/pull/46911#discussion_r1710982837 explains that why `local_idx` is different from `bind_index`.

Reply:
Rename to `local_idx` or `dag_idx`: https://github.com/ray-project/ray/pull/46911/commits/6e23a3ab1e5985ae98532d9cabe7d145e5830af3

Reply:
Rename: https://github.com/ray-project/ray/pull/46911/commits/5691e8b5e92d5a55fd9d4dde99d7d30176f51f9b

Reply:
@stephanie-wang is this example clear for you https://github.com/ray-project/ray/pull/46911#discussion_r1712192994? Thanks!

Reply:
Open an issue to track the naming of `idx` for the rest of the ADAG codebase: https://github.com/ray-project/ray/issues/47052